### DISCUSSION TITLE
The program was stopped (SystemAbort).

### DISCUSSION THREAD

**Discussion by j-schumacher** [pybricks/support#949] [Q&A] [answered]
I am a first time builder that could use some help with code to motorizing the Lego model for Ecto-1. The Lego Move Hub is being used with the two motors for the drive function and a separate Lego Large Motor V46 for the steering. 

I was having difficulties getting the various angle functions to operate to calibrate after running until stall but using a work around by running the motor in short burst that seems to work. The Move Hub motors operate at a constant speed along with a power burst that all appear to work fine ... for a little while. 

I receive the message "The program was stopped (SystemAbort)" after a random amount of time that appears to be related to the number of buttons pressed but not normalized in time or amount of presses. The Move Hub will stay blue for second after the remote control first disconnects then the Move Hub powers off.

I can continue this pattern by connecting back to bluetooth and running the code - I don't need to reflash the firmware. 

```python
from pybricks.pupdevices import Motor, Remote
from pybricks.parameters import Port, Direction, Stop, Button
from pybricks.tools import wait

# Initialize the motors.
steer = Motor(Port.C)
front = Motor(Port.A)
rear = Motor(Port.B, Direction.COUNTERCLOCKWISE)

# Connect to the remote.
remote = Remote()

# driving with remote
while True:
    # Check which buttons are pressed.
    pressed = remote.buttons.pressed()

    # steer by running the motor
    if Button.RIGHT_MINUS in pressed:
        steer.run_time(300,120,Stop,wait=False)
    if Button.RIGHT_PLUS in pressed:
        steer.run_time(-300,120,Stop,wait=False)

    # Choose the drive speed based on the left controls.
    drive_speed = 0
    if Button.LEFT_PLUS in pressed:
        drive_speed = 500
    if Button.LEFT_MINUS in pressed:
        drive_speed = -500
    
    # speed boost
    if Button.LEFT in pressed:
        drive_speed = 1200
        
    # Apply the selected speed.
    front.run(drive_speed)
    rear.run(drive_speed)
```

Is there debug or more verbose error messaging that can be enabled? Could there be a buffer overflow from this bad coding? How best to troubleshoot this situation?

Also, is there a tutorial on getting the code built-in with the firmware? How can the code run simply turning on the Move Hub and remote? Many thank you for reading and any helpful information you can share here.

**Comment by dlech** [comment]
> I receive the message "The program was stopped (SystemAbort)"

If you press the button on the hub or the stop button in Pybricks Code while the program is running, it will stop the program with this message.

> Is there debug or more verbose error messaging that can be enabled? 

The Move hub has very limited memory, so there isn't anything more verbose. You can add your own `print` statements for debugging.

> Also, is there a tutorial on getting the code built-in with the firmware?

When you turn off the hub, the current program is saved to flash memory. Just press the green button on the hub to start the program again at any time (unless the program is already running, then pressing the green button stops the program).

**Comment by j-schumacher** [comment, marked as answer]
Thank you for your response.

>  I receive the message "The program was stopped (SystemAbort)"
>> If you press the button on the hub or the stop button in Pybricks Code while the program is running, it will stop the program with this message.

I am not pressing the button on the hub or the stop button in Pybricks Code while the program is running.

>Is there debug or more verbose error messaging that can be enabled?
>> The Move hub has very limited memory, so there isn't anything more verbose. You can add your own print statements for debugging.

Understood

>Also, is there a tutorial on getting the code built-in with the firmware?
>>When you turn off the hub, the current program is saved to flash memory. Just press the green button on the hub to start the program again at any time (unless the program is already running, then pressing the green button stops the program).

That doesn't seem to happen in this situation but I am getting this shutdown that might be causing my problem as I need to use Pybricks Code application every time to run the code. I can disconnect from the application bluetooth to continue to run but after a short period of time if the shutdown happens on me.

Any suggestions?

**Reply by dlech** [reply]
Is the hub turning off due to low battery (light is flashing orange)?

**Reply by j-schumacher** [reply]
The battery is fully power, fresh AAA batteries and brand new from Lego. The app reports battery okay, full green

**Reply by laurensvalk** [reply]
> That doesn't seem to happen in this situation but I am getting this shutdown that might be causing my problem as I need to use Pybricks Code application every time to run the code. I can disconnect from the application bluetooth to continue to run but after a short period of time if the shutdown happens on me.

Try this: Load the program via Pybricks Code as usual, and exit it right away by pressing the button and then turn the hub off.

When you turn the hub on again, your program should still be there.

**Reply by laurensvalk** [reply]
> I receive the message "The program was stopped (SystemAbort)"

Could this message still have been there from previous attempts to run the program? To be sure, right click the output window and choose `Clear`.

**Reply by j-schumacher** [reply]
This is not a mistaken error on my part of a stale message.

**Reply by j-schumacher** [reply]
> Is the hub turning off due to low battery (light is flashing orange)?

I turned the hub back on and noticed the yellow light flashing then replaced the batteries. A little upset on the batteries not lasting very long but that isn't my problem for your side. The Pybricks Code was still reporting battery as green, fully charged.

I think my problems are solved from small testing. I was able to reconnect the remote with Hub directly. My only unresolved issue is with trying to set motor angle gives an error. I will seek to put in a buy report, thank you for helping here!

**Reply by laurensvalk** [reply]
It's a good point, though. Maybe we need to revisit the battery threshold levels to make it easier to see that the batteries are nearly empty.

**Comment by laurensvalk** [comment]
Does this still happen if you replace 

```python
steer.run_time(300,120,Stop,wait=False)
```

by

```python
steer.run_time(300,120,Stop.COAST,wait=False)
```

for both occurrences in your program?

Using a bare `Stop` is not supposed to work/allowed. This is a bug in the firmware; it is supposed to raise a `TypeError` for this. I've filed a bug at https://github.com/pybricks/support/issues/950. Thanks for your sample code.

**Reply by j-schumacher** [reply]
This didn't have an impact on the code but good to know and will keep at coast going forward here
